### PR TITLE
Yarn- Enhancements to Artifactory Integration and auto-installation for yarn projects

### PR DESCRIPTION
**Pull request by eranturgeman**
- [x] All [tests](https://github.com/jfrog/jfrog-cli-core#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-cli-core/actions/workflows/analysis.yml) passed.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

Summery of changes:

Implemented the ability to configure an Artifactory server as a resolution server while constructing the yarn dependency tree.

Introduced auto-install functionality for yarn projects within the process of building dependency trees. Users can also supply customized installation commands. Note that not all entry points support the provision of these parameters.

Resolved a bug related to the restoration of a backup file.

Added tests and test cases for the newly introduced functionalities.

### CODE REVIEW

File: xray/commands/audit/sca/yarn/yarn.go [hunk 1]
```diff
@@ -1,16 +1,40 @@
...
+	v1NonInteractiveFlag = "--non-interactive"
+	// Skips linking and fetch only packages that are missing from yarn.lock file
+	v2UpdateLockfileFlag = "--mode=update-lockfile"
+	// Ignores any build scripts
+	v2SkipBuildFlag     = "--mode=skip-build"
+	yarnV2Version       = "2.0.0"
+	nodeModulesRepoName = "node_modules"
+)
+
+func BuildDependencyTree(params utils.AuditParams) (dependencyTrees []*xrayUtils.GraphNode, uniqueDeps []string, err error) {
```
Comment (omerzi) [inline xray/commands/audit/sca/yarn/yarn.go:37]: Structs pass by value, make this change in other functions if needed
```suggestion
func BuildDependencyTree(params *utils.AuditParams) (dependencyTrees []*xrayUtils.GraphNode, uniqueDeps []string, err error) {
```

File: xray/commands/audit/sca/yarn/yarn.go [hunk 2]
```diff
@@ -24,6 +48,28 @@ func BuildDependencyTree() (dependencyTrees []*xrayUtils.GraphNode, uniqueDeps [
...
+	if !projectInstalled || len(params.InstallCommandArgs()) != 0 {
+		// In case project is not "installed" or in case the user has provided an 'install' command to run
+		var serverDetails *config.ServerDetails
+		serverDetails, err = params.ServerDetails()
+		if err != nil {
+			err = fmt.Errorf("failed to get server details while building yarn dependency tree: %s", err.Error())
+			return
+		}
+		depsRepo := params.DepsRepo()
+		err = configureYarnResolutionServerAndRunInstall(params, currentDir, executablePath, serverDetails, depsRepo)
```
Comment (omerzi) [inline xray/commands/audit/sca/yarn/yarn.go:66]: depsRepo and serverDetails exist in the `AuditParams`, so you don't need to pass them as parameters.

File: xray/commands/audit/sca/yarn/yarn.go [hunk 3]
```diff
@@ -35,6 +81,96 @@ func BuildDependencyTree() (dependencyTrees []*xrayUtils.GraphNode, uniqueDeps [
 	return
 }
 
+// Sets up Artifactory server configurations for dependency resolution, if such were provided by the user. Executes the user's 'install' command or a default 'install' command if none was specified.
```
Comment (omerzi) [inline xray/commands/audit/sca/yarn/yarn.go:84]:
```suggestion
// Sets up Artifactory server configurations for dependency resolution, if such were provided by the user. 
// Executes the user's 'install' command or a default 'install' command if none was specified.
```

File: xray/commands/audit/sca/yarn/yarn.go [hunk 4]
```diff
@@ -35,6 +81,96 @@ func BuildDependencyTree() (dependencyTrees []*xrayUtils.GraphNode, uniqueDeps [
...
+	// If the installCommandArgs in the parames is not empty, it signifies that the user has provided it, and 'install' is already included as one of the arguments
+	installCommandProvidedFromUser := len(installCommandArgs) != 0
+	if !installCommandProvidedFromUser {
+		installCommandArgs = []string{"install"}
+	}
+
+	// Upon receiving a user-provided 'install' command, we execute the command exactly as provided
+	if installCommandProvidedFromUser {
+		return build.RunYarnCommand(yarnExecPath, curWd, installCommandArgs...)
+	}
```
Comment (omerzi) [inline xray/commands/audit/sca/yarn/yarn.go:142]:
```suggestion
    // Upon receiving a user-provided 'install' command, we execute the command exactly as provided
	installCommandProvidedFromUser := len(installCommandArgs) != 0
	if installCommandProvidedFromUser {
		return build.RunYarnCommand(yarnExecPath, curWd, installCommandArgs...)
	}
    
    // If no installation command has been provided, we will execute 'yarn install'.
    installCommandArgs = []string{"install"}
```

File: xray/commands/audit/sca/yarn/yarn.go [hunk 5]
```diff
@@ -35,6 +81,96 @@ func BuildDependencyTree() (dependencyTrees []*xrayUtils.GraphNode, uniqueDeps [
...
+		nodeModulesDirExists, err = fileutils.IsDirExists(nodeModulesFullPath, false)
+		if err != nil {
+			err = fmt.Errorf("failed while checking for existence of node_modules directory: %s", err.Error())
+			return
+		}
+		if !nodeModulesDirExists {
+			defer func() {
+				err = errors.Join(err, fileutils.RemoveTempDir(nodeModulesFullPath))
+			}()
+		}
```
Comment (omerzi) [inline xray/commands/audit/sca/yarn/yarn.go:165]: Let's move the main logic that searches for 'node_modules' and deletes it if it exists into a dedicated function. Then, let's invoke this function using defer.

**Review by omerzi — CHANGES_REQUESTED** [review]

File: xray/commands/audit/sca/yarn/yarn.go (hunk 5 above)
Comment (eranturgeman) [reply xray/commands/audit/sca/yarn/yarn.go:165]: this will not work the way I intended it to work
I reviewed it again and kept it this way

**Pull request merged**